File: src/main/java/com/aoranzhang/ezrentback/data/entity/Building.java
package com.aoranzhang.ezrentback.data.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.validator.constraints.NotEmpty;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "building")
public class Building extends AbstractPersistentObject {
    @NotEmpty
    private String address;

    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @JoinColumn(name = "city", nullable = false)
    private City city;

    private String postalCode;

    @OneToMany(
            mappedBy = "building",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.EAGER
    )
    @JsonIgnore
    private Set<Suite> suites = new HashSet<>();

    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @JoinColumn(name = "owner", nullable = false)
    private User owner;

    private double longitude, latitude;

    private boolean verified;

    public double getLongitude() {
        return longitude;
    }

    public void setLongitude(double longitude) {
        this.longitude = longitude;
    }

    public double getLatitude() {
        return latitude;
    }

    public void setLatitude(double latitude) {
        this.latitude = latitude;
    }

    public boolean isVerified() {
        return verified;
    }

    public void setVerified(boolean verified) {
        this.verified = verified;
    }

    public User getOwner() {
        return owner;
    }

    public void setOwner(User owner) {
        this.owner = owner;
    }

    public void addSuite(Suite suite) {
        suite.setBuilding(this);
        suites.add(suite);
    }

    public void removeSuite(Suite suite) {
        suite.setBuilding(null);
        suites.remove(suite);
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public City getCity() {
        return city;
    }

    public void setCity(City city) {
        this.city = city;
    }

    public String getPostalCode() {
        return postalCode;
    }

    public void setPostalCode(String postalCode) {
        this.postalCode = postalCode;
    }

    public Set<Suite> getSuites() {
        return suites;
    }

    public void setSuites(Set<Suite> suites) {
        this.suites = suites;
    }
}

File: src/main/kotlin/com/aoranzhang/ezrentback/data/entity/Building.kt
package com.aoranzhang.ezrentback.data.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import com.google.common.collect.ImmutableSet

import javax.persistence.*
import java.util.HashSet

@Entity
@Table(name = "building")
class Building : AbstractPersistentObject {
    @ManyToOne(fetch = FetchType.EAGER, cascade = [CascadeType.ALL])
    @JoinColumn(name = "city", nullable = false)
    var city: City? = null

    var address:String = ""

    var postalCode: String = ""

    @OneToMany(mappedBy = "building", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonIgnore
    var suites: Set<Suite> = HashSet()

    @ManyToOne(fetch = FetchType.EAGER, cascade = [CascadeType.ALL])
    @JoinColumn(name = "owner", nullable = false)
    var owner: User? = null

    var longitude: Double = 0.0

    var latitude: Double = 0.0

    var verified: Boolean = false

    constructor() : super() {}

    constructor(city: City, address:String, postalCode: String, suites: Set<Suite>, owner: User, longitude: Double, latitude: Double, verified: Boolean) : super() {
        this.city = city
        this.address = address
        this.postalCode = postalCode
        this.suites = suites
        this.owner = owner
        this.latitude = latitude
        this.longitude = longitude
        this.verified = verified
    }


    constructor(building: Building) : super(building.id, building.version) {
        this.city = building.city
        this.address = building.address
        this.postalCode = building.postalCode
        this.suites = building.suites
        this.owner = building.owner
        this.latitude = building.latitude
        this.longitude = building.longitude
        this.verified = building.verified
    }
}
